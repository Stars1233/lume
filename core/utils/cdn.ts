import {
  canParse,
  format,
  maxSatisfying,
  parse,
  parseRange,
} from "../../deps/semver.ts";
import { globToRegExp } from "../../deps/path.ts";

type PackageType = "npm" | "gh";

export async function getFiles(
  specifier: string,
): Promise<Map<string, string>> {
  const result = parseNpm(specifier) || parseGh(specifier);
  if (!result) {
    throw new Error(`Invalid specifier: ${specifier}`);
  }
  const [type, name, range, pattern] = result;
  const version = await getVersion(type, name, range);
  if (!version) {
    throw new Error(`No matching version found for ${name}@${range}`);
  }
  const url =
    `https://data.jsdelivr.com/v1/package/${type}/${name}@${version}?structure=flat`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch package data for ${name}@${version}`);
  }
  const data = await res.json();
  const regexp = globToRegExp(pattern, { globstar: true, extended: true });
  const basePath = pattern.includes("*") ? pattern.split("/*")[0] : undefined;
  const fileMap = new Map<string, string>();

  for (const file of data.files) {
    const filename: string = file.name;

    if (!regexp.test(filename)) {
      continue;
    }

    const key = basePath ? filename.slice(basePath.length) : filename;
    fileMap.set(
      key,
      `https://cdn.jsdelivr.net/${type}/${name}@${version}${filename}`,
    );
  }

  return fileMap;
}

export function isFromCdn(specifier: string): boolean {
  return specifier.startsWith("npm:") || specifier.startsWith("gh:");
}

export function getFile(specifier: string): string {
  const result = parseNpm(specifier) || parseGh(specifier);
  if (!result) {
    throw new Error(`Invalid specifier: ${specifier}`);
  }

  const [type, name, version, filename] = result;

  if (filename.includes("*")) {
    throw new Error(`Specifier must not contain glob pattern: ${specifier}`);
  }

  return `https://cdn.jsdelivr.net/${type}/${name}@${version}${filename}`;
}

export async function getVersion(
  type: PackageType,
  name: string,
  version: string,
): Promise<string | undefined> {
  const url = `https://data.jsdelivr.com/v1/package/${type}/${name}`;
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Failed to fetch latest version for ${name}`);
  }
  const data = await res.json();

  const versions = data.versions.filter(canParse).map(parse);
  const range = parseRange(version);
  const found = maxSatisfying(versions, range);
  return found ? format(found) : undefined;
}

const NPM = /^npm:(@[^/]+\/)?([^@/]+)(@([^/]+))?(\/.+)?$/;
function parseNpm(
  specifier: string,
): [PackageType, string, string, string] | undefined {
  const match = specifier.match(NPM);

  if (!match) {
    return;
  }

  const [, scope, name, , version, pattern] = match;
  return [
    "npm",
    (scope || "") + name,
    version || "*",
    pattern || "/**",
  ];
}

const GITHUB = /^gh:([^/]+\/[^/@]+)(@([^/]+))?(\/.+)?$/;
function parseGh(
  specifier: string,
): [PackageType, string, string, string] | undefined {
  const match = specifier.match(GITHUB);

  if (!match) {
    return;
  }

  const [, name, , version, pattern] = match;
  return [
    "gh",
    name,
    version || "*",
    pattern || "/**",
  ];
}
