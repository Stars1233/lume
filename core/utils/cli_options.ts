import { parseArgs } from "../../deps/cli.ts";
import { getFreePort } from "./net.ts";
import { env } from "./env.ts";

import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port", "hostname"],
    boolean: ["serve", "open"],
    alias: { serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  // Detect location, hostname and port
  let port: number;
  let hostname: string;
  let location: URL;

  if (serveMode) {
    const proxied = env<boolean>("LUME_PROXIED");
    location = cli.location
      ? new URL(cli.location)
      : proxied
      ? (options.location as URL | undefined) || new URL("http://localhost")
      : new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      if (!proxied) {
        location.port = port.toString();
      }
    } else if (location.port && !proxied) {
      port = parseInt(location.port);
    } else if (options.server?.port) {
      port = options.server.port;
      if (!proxied) {
        location.port = port.toString();
      }
    } else if (!proxied) {
      port = location.protocol === "https:" ? 443 : getFreePort(3000, 3010);
      location.port = port.toString();
    } else {
      port = 8000;
    }

    if (cli.hostname) {
      hostname = cli.hostname;
      if (!proxied) {
        location.hostname = hostname;
      }
    } else if (options.server?.hostname && !cli.location && !proxied) {
      hostname = options.server.hostname;
      location.hostname = hostname;
    } else {
      hostname = proxied ? "localhost" : location.hostname;
    }
  } else {
    location = cli.location
      ? new URL(cli.location)
      : (options.location as URL | undefined) || new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      location.port = port.toString();
    } else if (location.port) {
      port = parseInt(location.port);
    } else {
      port = location.protocol === "https:" ? 443 : 80;
    }

    if (cli.hostname) {
      hostname = cli.hostname;
      location.hostname = hostname;
    } else {
      hostname = location.hostname;
    }
  }

  options.location = location;
  options.server ||= {};
  options.server.port = port;
  options.server.hostname = hostname;

  if (cli.open) {
    options.server.open = cli.open;
  }

  return options;
}
