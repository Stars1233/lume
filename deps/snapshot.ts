export * from "jsr:@std/testing@1.0.16/snapshot";
