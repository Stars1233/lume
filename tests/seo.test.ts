import { build, getSite } from "./utils.ts";
import { assertSnapshot } from "../deps/snapshot.ts";
import seo from "../plugins/seo.ts";

Deno.test("SEO plugin", async (t) => {
  const site = getSite({
    src: "seo",
  });

  const result: { url: string; messages: string[] }[] = [];

  site.use(seo({
    output(reports) {
      for (const [url, messages] of reports.entries()) {
        result.push({
          url,
          messages: messages.map((msg) =>
            typeof msg === "string" ? msg : msg.msg
          ),
        });
      }
    },
  }));

  await build(site);
  result.sort((a, b) => a.url.localeCompare(b.url));
  await assertSnapshot(t, result);
});
