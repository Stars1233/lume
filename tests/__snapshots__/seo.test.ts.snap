export const snapshot = {};

snapshot[`SEO plugin 1`] = `
[
  {
    messages: [
      "DESCRIPTION_MISSING",
      "BODY_TOO_SHORT",
    ],
    url: "/",
  },
]
`;
